File: lib/utils/undock_initiator.dart
part of dock_spawn;

/** Listen for undock events */
typedef Dialog OnUndock(MouseEvent e, Point2 dragOffset);

/**
 * Listens for events on the [element] and notifies the [listener]
 * if an undock event has been invoked.  An undock event is invoked
 * when the user clicks on the event and drags is beyond the 
 * specified [thresholdPixels]
 */ 
class UndockInitiator {
  Element element;
  OnUndock listener;
  num thresholdPixels;
  Point2 dragStartPosition;
  
  StreamSubscription<MouseEvent> mouseDownHandler;
  StreamSubscription<MouseEvent> mouseUpHandler;
  StreamSubscription<MouseEvent> mouseMoveHandler;
  
  bool _enabled = false;
  bool get enabled {
    return _enabled;
  }
  
  void set enabled(bool value) {
    _enabled = value;
    if (_enabled) {
      _cancelSubscription(mouseDownHandler);
      mouseDownHandler = element.onMouseDown.listen(onMouseDown);
    } else {
      _cancelSubscription(mouseDownHandler);
      _cancelSubscription(mouseUpHandler);
      _cancelSubscription(mouseMoveHandler);
      mouseDownHandler = null;
      mouseUpHandler = null;
      mouseMoveHandler = null;
    }
  }
  
  UndockInitiator(this.element, this.listener, [this.thresholdPixels = 10]) {
  }
  
  void _cancelSubscription(StreamSubscription sub) {
    if (sub != null) {
      sub.cancel();
    }
  }
  
  void onMouseDown(MouseEvent e) {
    // Make sure we dont do this on floating dialogs
    if (enabled) {
      _cancelSubscription(mouseUpHandler);
      _cancelSubscription(mouseMoveHandler);
      mouseUpHandler = window.onMouseUp.listen(onMouseUp);
      mouseMoveHandler = window.onMouseMove.listen(onMouseMove);
      dragStartPosition = new Point2(e.page.x, e.page.y);
    }
  }
  void onMouseUp(MouseEvent e) {
    _cancelSubscription(mouseUpHandler);
    _cancelSubscription(mouseMoveHandler);
    mouseUpHandler = null;
    mouseMoveHandler = null;
  }
  void onMouseMove(MouseEvent e) {
    Point2 position = new Point2(e.page.x, e.page.y);
    num dx = position.x - dragStartPosition.x;
    num dy = position.y - dragStartPosition.y;
    num distance = sqrt(dx * dx + dy * dy);
    
    if (distance > thresholdPixels) {
      enabled = false;
      _requestUndock(e);
    }
  }
  
  void _requestUndock(MouseEvent e) {
    num dragOffsetX = dragStartPosition.x - element.offset.left;
    num dragOffsetY = dragStartPosition.y - element.offset.top;
    Point2 dragOffset = new Point2(dragOffsetX, dragOffsetY);
    listener(e, dragOffset);
  }
}

File: js/lib/utils/UndockInitiator.js
/**
 * Listens for events on the [element] and notifies the [listener]
 * if an undock event has been invoked.  An undock event is invoked
 * when the user clicks on the event and drags is beyond the
 * specified [thresholdPixels]
 */
dockspawn.UndockInitiator = function(element, listener, thresholdPixels)
{
    if (!thresholdPixels)
        thresholdPixels = 10;

    this.element = element;
    this.listener = listener;
    this.thresholdPixels = thresholdPixels;
    this._enabled = false;
};

Object.defineProperty(dockspawn.UndockInitiator.prototype, "enabled", {
    get: function() { return this._enabled; },
    set: function(value)
    {
        this._enabled = value;
        if (this._enabled)
        {
            if (this.mouseDownHandler)
            {
                this.mouseDownHandler.cancel();
                delete this.mouseDownHandler;
            }
            this.mouseDownHandler = new dockspawn.EventHandler(this.element, 'mousedown', this.onMouseDown.bind(this));
        }
        else
        {
            if (this.mouseDownHandler)
            {
                this.mouseDownHandler.cancel();
                delete this.mouseDownHandler;
            }
            if (this.mouseUpHandler)
            {
                this.mouseUpHandler.cancel();
                delete this.mouseUpHandler;
            }
            if (this.mouseMoveHandler)
            {
                this.mouseMoveHandler.cancel();
                delete this.mouseMoveHandler;
            }
        }
    }
});

dockspawn.UndockInitiator.prototype.onMouseDown = function(e)
{
    // Make sure we dont do this on floating dialogs
    if (this.enabled)
    {
        if (this.mouseUpHandler)
        {
            this.mouseUpHandler.cancel();
            delete this.mouseUpHandler;
        }
        if (this.mouseMoveHandler)
        {
            this.mouseMoveHandler.cancel();
            delete this.mouseMoveHandler;
        }
        this.mouseUpHandler = new dockspawn.EventHandler(window, 'mouseup', this.onMouseUp.bind(this));
        this.mouseMoveHandler = new dockspawn.EventHandler(window, 'mousemove', this.onMouseMove.bind(this));
        this.dragStartPosition = new Point(e.pageX, e.pageY);
    }
};

dockspawn.UndockInitiator.prototype.onMouseUp = function(e)
{
    if (this.mouseUpHandler)
    {
        this.mouseUpHandler.cancel();
        delete this.mouseUpHandler;
    }
    if (this.mouseMoveHandler)
    {
        this.mouseMoveHandler.cancel();
        delete this.mouseMoveHandler;
    }
};

dockspawn.UndockInitiator.prototype.onMouseMove = function(e)
{
    var position = new Point(e.pageX, e.pageY);
    var dx = position.x - this.dragStartPosition.x;
    var dy = position.y - this.dragStartPosition.y;
    var distance = Math.sqrt(dx * dx + dy * dy);

    if (distance > this.thresholdPixels)
    {
        this.enabled = false;
        this._requestUndock(e);
    }
};

dockspawn.UndockInitiator.prototype._requestUndock = function(e)
{
    var dragOffsetX = this.dragStartPosition.x - this.element.offsetLeft;
    var dragOffsetY = this.dragStartPosition.y - this.element.offsetTop;
    var dragOffset = new Point(dragOffsetX, dragOffsetY);
    this.listener(e, dragOffset);
};
